parlai/core/pytorch_data_teacher.py

```diff
@@ -426,15 +426,32 @@ def __init__(self, opt):
         self.datapath = build_data(self.opt)
         self.data_gen = self._data_generator()
         self.length_datafile = os.path.join(self.datapath, 'data_length')
+        self.char_index_file = os.path.join(self.datapath, 'char_index')
         self.datafile = os.path.join(self.datapath, 'data')
         self.training = self.datatype.startswith('train')
+        self.ordered = ('ordered' in self.datatype or
+                        ('stream' in self.datatype and not opt.get('shuffle')))
         self._load_lens()
 
     def __getitem__(self, index):
-        while True:
-            idx, ep = next(self.data_gen)
-            if idx == index:
-                return (index, ep)
+        if self.ordered or not self.training:
+            while True:
+                idx, ep = next(self.data_gen)
+                if idx == index:
+                    return (index, ep)
+        else:
+            episode = []
+            current_idx = index
+            episode_done = False
+            with open(self.datafile) as f:
+                while not episode_done:
+                    ex_offset = self.char_index[current_idx]
+                    f.seek(ex_offset)
```

Comment:
the `f.seek` call should only be necessary for the first example in the episode, right? Otherwise they're sequential, so no seek is necessary? That would slightly speed it up.

Reply:
yes, that is a good point - I rewrote to simplify the code, but it did not actually result in any speedup for me